core/src/main/scala/codesearch/core/search/Search.scala

```diff
@@ -19,29 +19,37 @@ import io.chrisdavenport.log4cats.SelfAwareStructuredLogger
 import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
 
 import scala.sys.process.{Process, ProcessLogger}
+import com.google.re2j._
+
+import scala.util.{Success, Try}
 
 trait Search {
 
   protected def cindexDir: СindexDirectory
   protected def extensions: Extensions
   protected val logger: SelfAwareStructuredLogger[IO] = Slf4jLogger.unsafeCreate[IO]
 
+  def checkRegexpForValid(regexp: String): Try[Pattern] = {
+    Try(Pattern.compile(regexp))
+  }
+
   def search(request: SearchRequest): IO[CSearchPage] = {
-    val entity = csearch(request)
-    if (entity.error.message.isEmpty) {
-      for {
-        results <- Stream
-          .emits(entity.lists)
-          .through(SnippetsGrouper.groupLines(snippetConfig))
-          .drop(snippetConfig.pageSize * (request.page - 1))
-          .take(snippetConfig.pageSize)
-          .evalMap(createSnippet)
-          .through(groupByPackage)
-          .compile
-          .toList
-      } yield CSearchPage(results.sortBy(_.pack.name), entity.lists.size, ErrorResponse(""))
-    } else {
-      IO(CSearchPage(Seq.empty[Search.PackageResult], 0, entity.error))
+    checkRegexpForValid(request.query) match {
```

Comment:
If function is large, then it becomes hard to understand, so I think that should be separated into several smaller functions.